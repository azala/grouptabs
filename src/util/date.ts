export default {
  /**
   * Parses a date string in the format `yyyy-MM-dd` as being used by `<input type="date">`
   * consistently across different hosts (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse).
   */
  parseDate: function (input: string) {
    // backwards compatibility: strip time info from format being used earlier
    input = input.substring(0, 10);

    var params = input.split("-").map(function (x) {
      return parseInt(x, 10);
    });
    return new Date(params[0], params[1] - 1, params[2]);
  },

  /**
   * Formats a date to the format `yyyy-MM-dd` as being used by `<input type="date">`
   */
  formatDate: function (date: Date) {
    var month: number | string = date.getMonth() + 1;
    month = month < 10 ? "0" + month : month;
    var day: string | number = date.getDate();
    day = day < 10 ? "0" + day : day;
    return "" + date.getFullYear() + "-" + month + "-" + day;
  },

  formatHumanDate: function (date: Date | string) {
    if (typeof date === "string") {
      date = this.parseDate(date);
    }

    if (this.isToday(date)) {
      return "Today";
    }

    if (this.isYesterday(date)) {
      return "Yesterday";
    }

    return date.toLocaleDateString();
  },

  addDays: function (date: Date, days: number) {
    var ms = date.getTime();
    var daysInMs = days * 1000 * 60 * 60 * 24;
    return new Date(ms + daysInMs);
  },

  isSameDay: function (date1: Date, date2: Date) {
    return this.formatDate(date1) === this.formatDate(date2);
  },

  isToday: function (date: Date | string) {
    if (typeof date === "string") {
      date = this.parseDate(date);
    }

    var today = new Date();
    return this.isSameDay(today, date);
  },

  isYesterday: function (date: Date | string) {
    if (typeof date === "string") {
      date = this.parseDate(date);
    }

    var yesterday = this.addDays(new Date(), -1);
    return this.isSameDay(yesterday, date);
  },
};
