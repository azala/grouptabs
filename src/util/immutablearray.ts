// immutable array helpers

export default class ImmutableArrayHelper {
  static add(array: any[], item: any): any[] {
    return array.concat(item);
  }

  static addUniq(array: any[], item: any): any[] {
    var index = array.indexOf(item);
    if (index !== -1) {
      return array;
    }

    return ImmutableArrayHelper.add(array, item);
  }

  static remove(array: any[], index: number): any[] {
    array = array.slice();  // copy
    array.splice(index, 1);
    return array;
  }

  static removeItem(array: any[], item: any): any[] {
    var index = array.indexOf(item);

    if (index !== -1) {
      return ImmutableArrayHelper.remove(array, index);
    }

    return array;
  }
}
