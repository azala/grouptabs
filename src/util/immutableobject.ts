// immutable object helpers

type dict = {[key: string]: any};

export default class ImmutableObjectHelper {
  static merge(...objects: dict[]): dict {
    var result: dict = {};

    for (var i = 0; i < objects.length; i++) {
      var toBeMerged = objects[i];

      for (var key in toBeMerged) {
        result[key] = toBeMerged[key];
      }
    }

    return result;
  }

  static set(object: dict, key: string, data: any): dict {
    object = ImmutableObjectHelper.merge(object);  // copy
    object[key] = data;
    return object;
  }

  static remove(object: dict, key: string): dict {
    object = ImmutableObjectHelper.merge(object);  // copy
    delete object[key];
    return object;
  }
}
