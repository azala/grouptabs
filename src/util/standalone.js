// Safari on iOS has a "standalone" feature, aka "add to homescreen".
// Every time the user opens this app, it restarts from the URL
// that was originally set as the bookmark.
// The experience we want is that it starts from where the user navigated to last.
// In other similar environments, e.g. Chrome on Android add to homescreen
// the app is not restarted when re-opened.

var isStandaloneSafari = !!window.navigator.standalone && window.navigator.userAgent.indexOf('Safari') >= 0;
var storageKey = 'grouptabs_last_visited_location';

export function restoreLocation () {
  if (!isStandaloneSafari) {
    return;
  }

  var lastVisitedUrl = window.localStorage.getItem(storageKey);
  if (lastVisitedUrl) {
    window.location.replace(lastVisitedUrl);
  }
}

export function startPersistingLocation (history) {
  if (!isStandaloneSafari) {
    return;
  }

  history.listen(function () {
    window.localStorage.setItem(storageKey, window.location.href);
  });
}
