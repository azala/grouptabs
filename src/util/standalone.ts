// Safari on iOS has a "standalone" feature, aka "add to homescreen".
// Every time the user opens this app, it restarts from the URL
// that was originally set as the bookmark.
// The experience we want is that it starts from where the user navigated to last.
// In other similar environments, e.g. Chrome on Android add to homescreen
// the app is not always restarted when re-opened but will get killed after a while as well.

const isStandalone = !!(window.navigator as any).standalone || window.matchMedia('(display-mode: standalone)').matches || document.referrer.includes('android-app://');
const storageKey = 'grouptabs_last_visited_location';

export function restoreLocation () {
  if (!isStandalone) {
    return;
  }

  const lastVisitedUrl = window.localStorage.getItem(storageKey);
  if (lastVisitedUrl) {
    window.location.replace(lastVisitedUrl);
  }
}

export function startPersistingLocation (history: any) {
  if (!isStandalone) {
    return;
  }

  history.listen(function () {
    window.localStorage.setItem(storageKey, window.location.href);
  });
}
