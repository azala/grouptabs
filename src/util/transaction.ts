import { Transaction, TransactionType } from "../types";

export default {

  /**
   * Determines transaction type from transaction data.
   *
   * Old transaction data might not have `transactionType` specified as "DIRECT"
   * but semantically still be a direct transaction.
   */
  getTransactionType: function (transaction: Transaction): TransactionType {
    if (transaction.transactionType === TransactionType.DIRECT) {
      return transaction.transactionType;
    }

    // find out if it's a transaction with two participants where one paid amount X
    // and the other paid amount -X
    if (transaction.participants.length === 2) {
      var amounts = transaction.participants.map(function (participant) {
        return participant.amount;
      });
      if (amounts[0] && amounts[0] === -amounts[1]) {
        return TransactionType.DIRECT;
      }
    }

    return TransactionType.SHARED;
  }

};
