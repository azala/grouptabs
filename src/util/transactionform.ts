import {
  TransactionFormState,
  Account,
  Transaction,
  TransactionFormParticipantInputType as InputType,
  TransactionFormParticipantStatus as Status,
  TransactionType,
  DocumentType,
  TransactionFormSharedState,
} from "../types";
import { v4 as uuidv4 } from "uuid";
// @ts-ignore
import orderBy from "lodash.orderby";
import { formatDate, parseDate } from "./date";

export const NEW_PARTICIPANT_OPTION = "NEW_PARTICIPANT";

export function createNewParticipant() {
  return {
    id: uuidv4(),
    inputType: InputType.NEW,
    status: Status.JOINED,
  };
}

function createParticipantInputDataForEmptyTab(): TransactionFormState["shared"] {
  return [1, 2].map(() => createNewParticipant());
}

function createParticipantInputData(
  accounts: Account[],
  transaction?: Transaction
): TransactionFormState["shared"] {
  if (!accounts.length) {
    return createParticipantInputDataForEmptyTab();
  }

  const participantInputs: TransactionFormState["shared"] = accounts.map(
    (account) => {
      let status;
      let amount;
      if (transaction) {
        const value = transaction.participants.find((participantValue) => {
          return participantValue.participant === account.participant;
        });
        status =
          value && value.amount > 0
            ? Status.PAID
            : value && value.amount === 0
            ? Status.JOINED
            : Status.NONE;
        if (value?.amount !== undefined && value.amount > 0) {
          amount = value.amount;
        }
      } else {
        // For exactly two people in the tab, always set JOINED
        status = accounts.length === 2 ? Status.JOINED : Status.NONE;
      }

      return {
        id: uuidv4(),
        inputType: InputType.EXISTING,
        participant: account.participant,
        status,
        amount,
      };
    }
  );

  // sort participants by 1) amount paid 2) joined 3) alphabetical
  return orderBy(
    participantInputs,
    ["status", "amount", "participant"],
    ["desc", "desc", "asc"]
  );
}

interface Inputs {
  from?: string;
  to?: string;
  amount?: number;
}

function transactionParticipants2Inputs(participants: Account[]): Inputs {
  var result: Inputs = {};

  for (var i = 0; i < participants.length; ++i) {
    var participant = participants[i];

    if (!result.amount) {
      result.amount = Math.abs(participant.amount);
    }

    if (!result.from && participant.amount > 0) {
      result.from = participant.participant;
    }

    if (!result.to && participant.amount < 0) {
      result.to = participant.participant;
    }
  }

  return result;
}

function createDirectInputData(
  accounts: Account[],
  transaction?: Transaction
): TransactionFormState["direct"] {
  var inputProps = transactionParticipants2Inputs(
    transaction?.participants || []
  );

  var mostNegativeParticipant = accounts[0]?.participant;
  var mostPositiveParticipant =
    accounts[accounts.length - 1] && accounts[accounts.length - 1].participant;

  var tabParticipants = accounts.map(function (account) {
    return account.participant;
  });
  var options = tabParticipants.concat(NEW_PARTICIPANT_OPTION);

  return {
    options,
    from: inputProps.from || mostNegativeParticipant || NEW_PARTICIPANT_OPTION,
    to: inputProps.to || mostPositiveParticipant || NEW_PARTICIPANT_OPTION,
    amount: inputProps.amount,
  };
}

export function createFormData(
  accounts: Account[],
  transaction?: Transaction
): TransactionFormState {
  const shared = createParticipantInputData(
    accounts,
    transaction?.transactionType === TransactionType.SHARED
      ? transaction
      : undefined
  );
  const direct = createDirectInputData(
    accounts,
    transaction?.transactionType === TransactionType.DIRECT
      ? transaction
      : undefined
  );
  const form: TransactionFormState = {
    transactionType: TransactionType.SHARED,
    date: formatDate(new Date()),
    shared,
    direct,
  };

  if (transaction) {
    form.transactionType = transaction.transactionType;
    form.date = formatDate(parseDate(transaction.date));
    form.description = transaction.description;
  }

  return form;
}

function mapParticipant(
  participant: TransactionFormSharedState
): Account | undefined {
  if (!participant.participant || participant.status === Status.NONE) {
    return;
  }

  return {
    participant: participant.participant,
    amount:
      participant.status === Status.PAID && participant.amount
        ? participant.amount
        : 0,
  };
}

function mapParticipantsOfSharedTransaction(
  participants: TransactionFormState["shared"]
): Account[] {
  return participants
    .map(mapParticipant)
    .filter((participant) => participant) as Account[];
}

function mapParticipantsOfDirectTransaction(
  direct: TransactionFormState["direct"]
): Account[] {
  const participants = [
    {
      participant:
        direct.from === NEW_PARTICIPANT_OPTION ? direct.fromNew : direct.from,
      amount: direct.amount,
    },
    {
      participant:
        direct.to === NEW_PARTICIPANT_OPTION ? direct.toNew : direct.to,
      amount: -(direct.amount as number),
    },
  ] as Account[];

  return participants;
}

export function mapFormDataToTransaction(
  formData: TransactionFormState,
  tabId: string,
  id?: string
): Transaction {
  let participants: Account[];
  if (formData.transactionType === TransactionType.SHARED) {
    participants = mapParticipantsOfSharedTransaction(formData.shared);
  } else {
    participants = mapParticipantsOfDirectTransaction(formData.direct);
  }

  return {
    id: id || uuidv4(),
    type: DocumentType.TRANSACTION,
    tabId,
    timestamp: new Date().toJSON(),
    transactionType: formData.transactionType,
    description: formData.description || "",
    date: formData.date,
    participants,
  };
}

function validateShared(participants: TransactionFormState["shared"]): boolean {
  var joinedParticipants = participants.filter(function (participant) {
    return participant.status > 0;
  });

  if (joinedParticipants.length < 2) {
    return false;
  }

  // every joined participant needs a name
  for (var i = 0; i < joinedParticipants.length; i++) {
    if (!joinedParticipants[i].participant) {
      return false;
    }
  }

  var payingParticipants = joinedParticipants.filter(function (participant) {
    return participant.status === Status.PAID && participant.amount;
  });
  if (!payingParticipants.length) {
    return false;
  }

  return true;
}

function validateDirect(data: TransactionFormState["direct"]): boolean {
  const from = data.from === NEW_PARTICIPANT_OPTION ? data.fromNew : data.from;
  const to = data.to === NEW_PARTICIPANT_OPTION ? data.toNew : data.to;
  return !!from && !!to && !!data.amount && from !== to;
}

export function validate(data: TransactionFormState): boolean {
  if (!data.description) {
    return false;
  }

  if (!data.date) {
    return false;
  }

  if (data.transactionType === TransactionType.DIRECT) {
    return validateDirect(data.direct);
  } else {
    return validateShared(data.shared);
  }
}
