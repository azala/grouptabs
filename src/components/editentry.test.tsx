import React from "react";
import renderer from "react-test-renderer";
import EditEntry from "./editentry";
import {
  TransactionType,
  TransactionFormParticipantStatus,
  TransactionFormParticipantInputType,
} from "../types";
import { createFormData } from "../util/transactionform";

let realDate: any;

beforeAll(() => {
  const currentDate = new Date("2020-02-20T20:20:20.202Z");
  realDate = Date;
  global.Date = class extends Date {
    constructor(date: any, ...args: any[]) {
      if (date) {
        // @ts-ignore
        return super(date, ...args);
      }

      return currentDate;
    }
  } as any;
});

afterAll(() => {
  global.Date = realDate;
});

it("renders loader when checking remote tab", () => {
  const tree = renderer
    .create(
      <EditEntry
        mode="new"
        checkingRemoteTab={true}
        onInitForm={jest.fn()}
        onUpdateForm={jest.fn()}
        onUpdateSharedForm={jest.fn()}
        onUpdateDirectForm={jest.fn()}
        onUpdateParticipant={jest.fn()}
        onAddParticipant={jest.fn()}
        onSetAllJoined={jest.fn()}
        onChangeTabClick={jest.fn()}
        onCloseClick={jest.fn()}
        onSave={jest.fn()}
        onDelete={jest.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders empty form", () => {
  const tree = renderer
    .create(
      <EditEntry
        mode="new"
        formState={createFormData([])}
        onInitForm={jest.fn()}
        onUpdateForm={jest.fn()}
        onUpdateSharedForm={jest.fn()}
        onUpdateDirectForm={jest.fn()}
        onUpdateParticipant={jest.fn()}
        onAddParticipant={jest.fn()}
        onSetAllJoined={jest.fn()}
        onChangeTabClick={jest.fn()}
        onCloseClick={jest.fn()}
        onSave={jest.fn()}
        onDelete={jest.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders loader in edit mode with no data", () => {
  const tree = renderer
    .create(
      <EditEntry
        mode="edit"
        onInitForm={jest.fn()}
        onUpdateForm={jest.fn()}
        onUpdateSharedForm={jest.fn()}
        onUpdateDirectForm={jest.fn()}
        onUpdateParticipant={jest.fn()}
        onAddParticipant={jest.fn()}
        onSetAllJoined={jest.fn()}
        onChangeTabClick={jest.fn()}
        onCloseClick={jest.fn()}
        onSave={jest.fn()}
        onDelete={jest.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders prefilled form", () => {
  const tree = renderer
    .create(
      <EditEntry
        mode="edit"
        formState={{
          description: "DESCRIPTION",
          transactionType: TransactionType.SHARED,
          date: "2020-03-24",
          shared: [
            {
              id: "1",
              inputType: TransactionFormParticipantInputType.EXISTING,
              participant: "Martin",
              status: TransactionFormParticipantStatus.JOINED,
              amount: 5,
            },
            {
              id: "2",
              inputType: TransactionFormParticipantInputType.EXISTING,
              participant: "Jan",
              status: TransactionFormParticipantStatus.JOINED,
              amount: 0,
            },
          ],
          direct: {
            options: [],
          },
        }}
        onInitForm={jest.fn()}
        onUpdateForm={jest.fn()}
        onUpdateSharedForm={jest.fn()}
        onUpdateDirectForm={jest.fn()}
        onUpdateParticipant={jest.fn()}
        onAddParticipant={jest.fn()}
        onSetAllJoined={jest.fn()}
        onChangeTabClick={jest.fn()}
        onCloseClick={jest.fn()}
        onSave={jest.fn()}
        onDelete={jest.fn()}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
