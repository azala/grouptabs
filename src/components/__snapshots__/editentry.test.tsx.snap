// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty form 1`] = `
<div
  class="scene editEntryScene"
>
  <div
    class="header"
  >
    <button
      class="left"
    >
      <svg
        height="16"
        width="16"
      >
        <path
          d="m7.4983 0.5c0.8974 0 1.3404 1.0909 0.6973 1.7168l-4.7837 4.7832h11.573c1.3523-0.019125 1.3523 2.0191 0 2h-11.572l4.7832 4.7832c0.98163 0.94251-0.47155 2.3957-1.4141 1.4141l-6.4911-6.49c-0.387-0.3878-0.391-1.0228 0-1.414l6.4905-6.49c0.1883-0.1935 0.4468-0.30268 0.7168-0.3027z"
        />
      </svg>
    </button>
    <h2>
      New transaction
    </h2>
    <button
      class="right create"
      form="edit-entry-form"
    >
      <svg
        height="16"
        width="16"
      >
        <path
          d="m13.631 3.9906a1.0001 1.0001 0 0 0-0.6875 0.30273l-6.5937 6.5938-3.293-3.293a1.0001 1.0001 0 1 0-1.4141 1.4141l4 4a1.0001 1.0001 0 0 0 1.4141 0l7.3008-7.3008a1.0001 1.0001 0 0 0-0.72656-1.7168z"
          fill="#fff"
        />
      </svg>
    </button>
  </div>
  <div>
    <form
      id="edit-entry-form"
    >
      <div
        class="form"
      >
        <div
          class="form-row"
        >
          <div
            class="form-row-input description"
          >
            <input
              placeholder="Description"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-row-input transaction-type"
          >
            <select>
              <option
                value="SHARED"
              >
                Shared
              </option>
              <option
                value="DIRECT"
              >
                Direct
              </option>
            </select>
          </div>
        </div>
        <div
          class="form-row"
        >
          <div
            class="form-row-input"
          >
            <div
              class="form-row-input date-input"
            >
              <button
                class="selected"
                type="button"
              >
                today
              </button>
              <button
                class=""
                type="button"
              >
                yesterday
              </button>
              <input
                type="date"
                value="2020-02-20"
              />
            </div>
          </div>
        </div>
        <div
          class="form-row"
          style="display: none;"
        >
          <div
            class="direct-transaction"
          >
            <div
              class="form-row"
              style="display: none;"
            >
              <div
                class="form-row-input"
              >
                <select
                  class="full-width"
                >
                  <option
                    value="NEW_PARTICIPANT"
                  >
                    New participant…
                  </option>
                </select>
              </div>
            </div>
            <div
              class="form-row"
            >
              <div
                class="form-row-input"
              >
                <input
                  placeholder="Name …"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="direct-transaction-amount"
            >
              <svg
                height="16"
                width="16"
              >
                <path
                  d="m15.511 8.5129c0-0.8974-1.0909-1.3404-1.7168-0.6973l-4.7832 4.7837v-11.573c0.019125-1.3523-2.0191-1.3523-2 0v11.572l-4.7832-4.7832c-0.94251-0.98163-2.3957 0.47155-1.4141 1.4141l6.49 6.4911c0.3878 0.387 1.0228 0.391 1.414 0l6.4903-6.4906c0.1935-0.1883 0.30268-0.4468 0.3027-0.7168z"
                />
              </svg>
              <input
                placeholder="0"
                step="any"
                type="number"
                value=""
              />
            </div>
            <div
              class="form-row"
              style="display: none;"
            >
              <div
                class="form-row-input"
              >
                <select
                  class="full-width"
                >
                  <option
                    value="NEW_PARTICIPANT"
                  >
                    New participant…
                  </option>
                </select>
              </div>
            </div>
            <div
              class="form-row"
            >
              <div
                class="form-row-input"
              >
                <input
                  placeholder="Name …"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="form-row"
          >
            <div
              class="form-row-input"
            >
              <div
                class="newParticipantInput selected"
              >
                <span
                  class="participant"
                >
                  <input
                    placeholder="Name …"
                    type="text"
                    value=""
                  />
                </span>
                <span
                  class="participationStatus"
                >
                  <span
                    class="joinedButtonWrapper"
                  >
                    <button
                      class=" selected"
                      type="button"
                    >
                      joined
                    </button>
                  </span>
                  <button
                    class="paid-button"
                    type="button"
                  >
                    paid
                  </button>
                  <span
                    class="amountInput hidden"
                  >
                    <input
                      disabled=""
                      placeholder="0"
                      step="any"
                      type="number"
                      value=""
                    />
                  </span>
                </span>
              </div>
              <div
                class="newParticipantInput selected"
              >
                <span
                  class="participant"
                >
                  <input
                    placeholder="Name …"
                    type="text"
                    value=""
                  />
                </span>
                <span
                  class="participationStatus"
                >
                  <span
                    class="joinedButtonWrapper"
                  >
                    <button
                      class=" selected"
                      type="button"
                    >
                      joined
                    </button>
                  </span>
                  <button
                    class="paid-button"
                    type="button"
                  >
                    paid
                  </button>
                  <span
                    class="amountInput hidden"
                  >
                    <input
                      disabled=""
                      placeholder="0"
                      step="any"
                      type="number"
                      value=""
                    />
                  </span>
                </span>
              </div>
            </div>
          </div>
          <div
            class="form-row"
          >
            <div
              class="form-row-input"
            >
              <button
                type="button"
              >
                + new participant
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      />
    </form>
  </div>
</div>
`;

exports[`renders loader in edit mode with no data 1`] = `
<div
  class="scene editEntryScene"
>
  <div
    class="header"
  >
    <button
      class="left"
    >
      <svg
        height="16"
        width="16"
      >
        <path
          d="m7.4983 0.5c0.8974 0 1.3404 1.0909 0.6973 1.7168l-4.7837 4.7832h11.573c1.3523-0.019125 1.3523 2.0191 0 2h-11.572l4.7832 4.7832c0.98163 0.94251-0.47155 2.3957-1.4141 1.4141l-6.4911-6.49c-0.387-0.3878-0.391-1.0228 0-1.414l6.4905-6.49c0.1883-0.1935 0.4468-0.30268 0.7168-0.3027z"
        />
      </svg>
    </button>
    <h2>
      Edit transaction
    </h2>
  </div>
  <div>
    <div
      class="load-error"
    >
      <img
        alt=""
        src="logo.png"
      />
      <h2>
        Grouptabs
      </h2>
      <p>
        Could not find transaction.
      </p>
      <button
        class="create"
      >
        OK
      </button>
    </div>
  </div>
</div>
`;

exports[`renders loader when checking remote tab 1`] = `
<div
  class="scene editEntryScene"
>
  <div
    class="header"
  >
    <button
      class="left"
    >
      <svg
        height="16"
        width="16"
      >
        <path
          d="m7.4983 0.5c0.8974 0 1.3404 1.0909 0.6973 1.7168l-4.7837 4.7832h11.573c1.3523-0.019125 1.3523 2.0191 0 2h-11.572l4.7832 4.7832c0.98163 0.94251-0.47155 2.3957-1.4141 1.4141l-6.4911-6.49c-0.387-0.3878-0.391-1.0228 0-1.414l6.4905-6.49c0.1883-0.1935 0.4468-0.30268 0.7168-0.3027z"
        />
      </svg>
    </button>
    <h2>
      New transaction
    </h2>
  </div>
  <div
    class="loader tab-loader"
  >
    <div />
  </div>
</div>
`;

exports[`renders prefilled form 1`] = `
<div
  class="scene editEntryScene"
>
  <div
    class="header"
  >
    <button
      class="left"
    >
      <svg
        height="16"
        width="16"
      >
        <path
          d="m7.4983 0.5c0.8974 0 1.3404 1.0909 0.6973 1.7168l-4.7837 4.7832h11.573c1.3523-0.019125 1.3523 2.0191 0 2h-11.572l4.7832 4.7832c0.98163 0.94251-0.47155 2.3957-1.4141 1.4141l-6.4911-6.49c-0.387-0.3878-0.391-1.0228 0-1.414l6.4905-6.49c0.1883-0.1935 0.4468-0.30268 0.7168-0.3027z"
        />
      </svg>
    </button>
    <h2>
      Edit transaction
    </h2>
    <button
      class="right create"
      form="edit-entry-form"
    >
      <svg
        height="16"
        width="16"
      >
        <path
          d="m13.631 3.9906a1.0001 1.0001 0 0 0-0.6875 0.30273l-6.5937 6.5938-3.293-3.293a1.0001 1.0001 0 1 0-1.4141 1.4141l4 4a1.0001 1.0001 0 0 0 1.4141 0l7.3008-7.3008a1.0001 1.0001 0 0 0-0.72656-1.7168z"
          fill="#fff"
        />
      </svg>
    </button>
  </div>
  <div>
    <form
      id="edit-entry-form"
    >
      <div
        class="form"
      >
        <div
          class="form-row"
        >
          <div
            class="form-row-input description"
          >
            <input
              placeholder="Description"
              type="text"
              value="DESCRIPTION"
            />
          </div>
          <div
            class="form-row-input transaction-type"
          >
            <select>
              <option
                value="SHARED"
              >
                Shared
              </option>
              <option
                value="DIRECT"
              >
                Direct
              </option>
            </select>
          </div>
        </div>
        <div
          class="form-row"
        >
          <div
            class="form-row-input"
          >
            <div
              class="form-row-input date-input"
            >
              <button
                class=""
                type="button"
              >
                today
              </button>
              <button
                class=""
                type="button"
              >
                yesterday
              </button>
              <input
                type="date"
                value="2020-03-24"
              />
            </div>
          </div>
        </div>
        <div
          class="form-row"
          style="display: none;"
        >
          <div
            class="direct-transaction"
          >
            <div
              class="form-row"
              style="display: none;"
            >
              <div
                class="form-row-input"
              >
                <select
                  class="full-width"
                />
              </div>
            </div>
            <div
              class="form-row"
              style="display: none;"
            >
              <div
                class="form-row-input"
              >
                <input
                  placeholder="Name …"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="direct-transaction-amount"
            >
              <svg
                height="16"
                width="16"
              >
                <path
                  d="m15.511 8.5129c0-0.8974-1.0909-1.3404-1.7168-0.6973l-4.7832 4.7837v-11.573c0.019125-1.3523-2.0191-1.3523-2 0v11.572l-4.7832-4.7832c-0.94251-0.98163-2.3957 0.47155-1.4141 1.4141l6.49 6.4911c0.3878 0.387 1.0228 0.391 1.414 0l6.4903-6.4906c0.1935-0.1883 0.30268-0.4468 0.3027-0.7168z"
                />
              </svg>
              <input
                placeholder="0"
                step="any"
                type="number"
                value=""
              />
            </div>
            <div
              class="form-row"
              style="display: none;"
            >
              <div
                class="form-row-input"
              >
                <select
                  class="full-width"
                />
              </div>
            </div>
            <div
              class="form-row"
              style="display: none;"
            >
              <div
                class="form-row-input"
              >
                <input
                  placeholder="Name …"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="form-row"
          >
            <div
              class="form-row-input"
            >
              <div
                class="participantInput selected"
              >
                <span
                  class="participant"
                >
                  Martin
                </span>
                <span
                  class="participationStatus"
                >
                  <span
                    class="joinedButtonWrapper"
                  >
                    <button
                      class=" selected"
                      type="button"
                    >
                      joined
                    </button>
                  </span>
                  <button
                    class="paid-button"
                    type="button"
                  >
                    paid
                  </button>
                  <span
                    class="amountInput hidden"
                  >
                    <input
                      disabled=""
                      placeholder="0"
                      step="any"
                      type="number"
                      value="5"
                    />
                  </span>
                </span>
              </div>
              <div
                class="participantInput selected"
              >
                <span
                  class="participant"
                >
                  Jan
                </span>
                <span
                  class="participationStatus"
                >
                  <span
                    class="joinedButtonWrapper"
                  >
                    <button
                      class=" selected"
                      type="button"
                    >
                      joined
                    </button>
                  </span>
                  <button
                    class="paid-button"
                    type="button"
                  >
                    paid
                  </button>
                  <span
                    class="amountInput hidden"
                  >
                    <input
                      disabled=""
                      placeholder="0"
                      step="any"
                      type="number"
                      value="0"
                    />
                  </span>
                </span>
              </div>
            </div>
          </div>
          <div
            class="form-row"
          >
            <div
              class="form-row-input"
            >
              <button
                type="button"
              >
                + new participant
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row button-row"
      >
        <span
          class="fake-link delete"
        >
          Delete transaction
        </span>
      </div>
    </form>
  </div>
</div>
`;
